apps/drec-api/src/pods/issuer/issuer.service.ts

```diff
@@ -0,0 +1,172 @@
+        return await this.issueCertificateForGroup(group, startDate, endDate);
+      }),
+    );
+    const devicesWithoutGroups = await this.deviceService.findMultiple({
+      where: {
+        groupId: null,
+      },
+    });
+
+    if (devicesWithoutGroups.length) {
+      const ownerGroupedDevices = this.groupBy(
+        devicesWithoutGroups,
+        deviceGroupRule1,
+      );
+
+      await Promise.all(
+        ownerGroupedDevices?.map(async (ownerBasedGroup: Device[], i) => {
+          this.groupBy(ownerBasedGroup, deviceGroupRule2)?.map(
+            async (countryBasedGroup: Device[], j) => {
+              const categorizedGroup: IDeviceGroup = {
+                id: 0,
+                name: `${ownerBasedGroup[i][deviceGroupRule1]}_${countryBasedGroup[j][deviceGroupRule2]}`,
+                organizationId: ownerBasedGroup[i].registrant_organisation_code,
+                devices: countryBasedGroup,
+              };
+              return await this.issueCertificateForGroup(
+                categorizedGroup,
+                startDate,
+                endDate,
+              );
+            },
+          );
+        }),
+      );
+    }
+  }
+
+  private async issueCertificateForGroup(
+    group: IDeviceGroup,
+    startDate: DateTime,
+    endDate: DateTime,
+  ): Promise<void> {
+    const readsFilter: FilterDTO = {
+      offset: 0,
+      limit: 1000,
+      start: startDate.toString(),
+      end: endDate.toString(),
+    };
+
+    const org = await this.organizationService.findOne(group.organizationId);
+    if (!org) {
+      throw new NotFoundException(
+        `No organization found with code ${group.organizationId}`,
+      );
+    }
+    const groupReads: number[] = [];
+    await Promise.all(
+      group.devices.map(async (device: Device) =>
+        groupReads.push(await this.getDeviceFullReads(device.id, readsFilter)),
+      ),
+    );
+    const totalReadValue = groupReads.reduce(
+      (accumulator, currentValue) => accumulator + currentValue,
+      0,
+    );
+    // Convert from W to kW
+    const totalReadValueKw = Math.round(totalReadValue * 10 ** -3);
+    console.log('totalReadValue: ', totalReadValue);
+    const issuance: IIssueCommandParams<ICertificateMetadata> = {
+      deviceId: group.id.toString(), // groupID
+      energyValue: totalReadValueKw.toString(),
+      fromTime: new Date(startDate.toString()),
+      toTime: new Date(endDate.toString()),
+      toAddress: org.blockchainAccountAddress,
+      userId: org.blockchainAccountAddress,
+      metadata: {
+        type: CertificateSourceType.Generator,
+        deviceIds: group.devices.map((device: Device) => device.id),
+        id: group.id.toString(),
+      },
+    };
+    this.logger.log(
+      `Issuance: ${JSON.stringify(issuance)}, Group name: ${group.name}`,
+    );
+    // return;
+    return await this.issueCertificate(issuance);
+  }
+
+  private async getDeviceFullReads(
```

Comment:
can't this function be replaced by (base) read service aggregate sum?

Reply:
I thought so too initially but looking over aggregate sum, it requires additional params like window and difference.